src/ssl.c

```diff
@@ -34433,6 +34480,117 @@ void *wolfSSL_OPENSSL_malloc(size_t a)
     return (void *)XMALLOC(a, NULL, DYNAMIC_TYPE_OPENSSL);
 }
 
+int wolfSSL_OPENSSL_hexchar2int(unsigned char c)
+{
+    int ret = -1;
+
+    switch (c) {
+        case '0':
+            ret = 0;
+            break;
+        case '1':
+            ret = 1;
+            break;
+        case '2':
+            ret = 2;
+            break;
+        case '3':
+            ret = 3;
+            break;
+        case '4':
+              ret = 4;
+              break;
+        case '5':
+              ret = 5;
+              break;
+        case '6':
+              ret = 6;
+              break;
+        case '7':
+              ret = 7;
+              break;
+        case '8':
+              ret = 8;
+              break;
+        case '9':
+              ret = 9;
+              break;
+        case 'a':
+        case 'A':
+              ret = 0x0A;
+              break;
+        case 'b':
+        case 'B':
+              ret = 0x0B;
+              break;
+        case 'c':
+        case 'C':
+              ret = 0x0C;
+              break;
+        case 'd':
+        case 'D':
+            ret = 0x0D;
+            break;
+        case 'e':
+        case 'E':
+            ret = 0x0E;
+            break;
+        case 'f':
+        case 'F':
+            ret = 0x0F;
+            break;
+        default:
+            break;
+    }
+
+    return ret;
+}
+
+unsigned char *wolfSSL_OPENSSL_hexstr2buf(const char *str, long *len)
+{
+    unsigned char* targetBuf;
+    int srcDigitHigh = 0;
+    int srcDigitLow = 0;
+    size_t srcLen;
+    size_t srcIdx = 0;
+    size_t targetIdx = 0;
+
+    srcLen = XSTRLEN(str);
+    targetBuf = (unsigned char*)XMALLOC(srcLen / 2, NULL, DYNAMIC_TYPE_OPENSSL);
+    if (targetBuf == NULL) {
+        return NULL;
+    }
+
+    while (srcIdx < srcLen) {
+        if (str[srcIdx] == ':') {
+            srcIdx++;
+            continue;
+        }
+
+        srcDigitHigh = wolfSSL_OPENSSL_hexchar2int(str[srcIdx++]);
+
+        /* Odd number of digits */
+        if (srcIdx == srcLen) {
```

Comment:
Do you need this?
String is NUL terminated.

Reply:
I don't follow why that's relevant for this check. Idea is we're grabbing digits in pairs (2 digits for each hex value), and if there's no second character for a pair, then the hex string is invalid. OpenSSL does a similar check in their version of the function. Perhaps you're suggesting a different way to do this check?

Reply:
You know there is one digit from the while check.
You know that the string is NUL terminated.
will be either <char><char> or <char><NUL>.
The conversion of the character to number will return -1 when it sees a NUL.
Don't copy OpenSSL. Write the implementation from scratch.

Reply:
Ah, ok now I see what you’re saying. Yeah, I had initially written this after looking at how OpenSSL did it, but about a month after that I wrote it from scratch without looking at my old implementation so that I wasn’t merely copying what OpenSSL does, as that seems like bad form. I only referenced how they did it here because that approach seemed reasonable to me, but your idea makes sense, too, now that I understand it.